lib/common/http2_client.go

```diff
@@ -2,71 +2,58 @@ package common
 
 import (
 	"bytes"
-	"context"
 	"crypto/tls"
 	"io"
 	"net"
 	"net/http"
-	"sync"
 	"time"
 
 	"golang.org/x/net/http2"
 )
 
 type HTTP2Client struct {
-	sync.Mutex
-
 	client http.Client
-	conn   net.Conn
+	transport *http.Transport
 }
 
 func NewHTTP2Client(timeout, idleTimeout time.Duration, keepAlive bool) (client *HTTP2Client, err error) {
 	if keepAlive {
 		timeout, idleTimeout = 0, 0
 	}
 
-	dialer := &net.Dialer{
-		Timeout:   1 * time.Second,
-		KeepAlive: 100000 * time.Second,
-		DualStack: true,
-	}
-
-	client = &HTTP2Client{}
-
 	transport := &http.Transport{
 		TLSClientConfig: &tls.Config{
 			InsecureSkipVerify: true,
 		},
 		IdleConnTimeout:   idleTimeout,
 		DisableKeepAlives: !keepAlive,
-		DialContext: func(ctx context.Context, network, addr string) (conn net.Conn, err error) {
-			conn, err = dialer.DialContext(ctx, network, addr)
-
-			client.conn = conn
-			return conn, err
-		},
+		DialContext: (&net.Dialer{
+			Timeout:   1 * time.Second,
+			KeepAlive: 100000 * time.Second,
+			DualStack: true,
+		}).DialContext,
 	}
 
 	if err = http2.ConfigureTransport(transport); err != nil {
 		return
 	}
 
-	client.client = http.Client{
-		Transport: transport,
-		Timeout:   timeout,
-		CheckRedirect: func(req *http.Request, via []*http.Request) error {
-			return http.ErrUseLastResponse // NOTE prevent redirect
+	client = &HTTP2Client{
+		client: http.Client{
+			Transport: transport,
+			Timeout:   timeout,
+			CheckRedirect: func(req *http.Request, via []*http.Request) error {
+				return http.ErrUseLastResponse // NOTE prevent redirect
+			},
 		},
+		transport: transport,
 	}
 
 	return
 }
 
 func (c *HTTP2Client) Close() {
-	if c.conn == nil {
-		return
-	}
-	c.conn.Close()
+	c.transport.CloseIdleConnections()
```

Comment:
When I wrote this code, I want to `Close` client connection at anytime even in use, not idle. Please reflect this requirement.

Reply:
The problem with this class is that a `Client` does not map to a single connection.
See [the documentation](https://golang.org/pkg/net/http/#Client):

> The Client's Transport typically has internal state (cached TCP connections), so Clients should be reused instead of created as needed. Clients are safe for concurrent use by multiple goroutines.

The `Client` should not mess with the connection, it's the `Transport` role.
As for cancelling a request, it should be done with `Context.Cancel` on the `Request`.
So nothing the `Client` should know about.

Reply:
@Geod24 Okay, I also already know the `Request.WithContext()` way, but I wanted to close in `net.Con`. Anyway you wrote this code to consider context way, it will be good.